### ISSUE TITLE
[azeventhubs] NextPartitionClient does not always respect context cancellation

### ISSUE THREAD

**Issue by ryepup** [Azure/azure-sdk-for-go#23124]
### Bug Report

- import path of package in question `github.com/Azure/azure-sdk-for-go/sdk/messaging/azeventhubs`
- SDK version `v1.0.2`
- output of `go version`: `go version go1.22.3 darwin/amd64`

- What happened?

`NextPartitionClient` only respects the `ctx` if `Run` was successful.

I called `azeventhubs.Processor.Run` in one goroutine, and `azeventhubs.Processor.NextPartitionClient` in another. My eventhub was misconfigured and `Run` was returning an "Ip has been prevented to connect to the endpoint" error.

When `Run` errors out, that canceled the `ctx` I passed to `NextPartitionClient`, but `NextPartitionClient` never exited. I needed to manually kill the process.

- What did you expect or want to happen?

I expected `NextPartitionClient` to exit when it's context was canceled, regardless of what has happened in `Run`.

- How can we reproduce it?

Create a processor and a context, cancel the context, then call `NextPartitiionClient`.  Per the doc comment,`NextPartitionClient` should return nil, but it hangs forever.

```go
processor, _ := azeventhubs.NewProcessor(...)
ctx, cancel := context.WithCancel(context.Background())
cancel()
_ = processor.NextPartitionClient(ctx)
```

For a more realistic repro, configure an eventhubs that does not allow your IP address, and use something like

```go
processor, _ := azeventhubs.NewProcessor(...)
g, ctx := errgroup.WithContext(context.Background())
g.Go(func() error { return processor.Run(ctx) })
g.Go(func() error {
	_ = processor.NextPartitionClient(ctx)
	return nil
})
g.Wait()
```

In that case, the `Run` call fails, which will cancel the `ctx`, but `NextPartitionClient` will never return.

**Comment by ryepup** [comment]
I think it's getting hung up on this channel https://github.com/Azure/azure-sdk-for-go/blob/227230a398ba66a92313a607080af03f970701f9/sdk/messaging/azeventhubs/processor.go#L196

Maybe that can be
```go
select {
case <-p.runCalled:
case <-ctx.Done():
	return nil
}
```

**Comment by richardpark-msft** [comment]
@ryepup, this bug was fixed as part of all the work done for v1.2.0: https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/messaging/azeventhubs/CHANGELOG.md#bugs-fixed-2

**Comment by ryepup** [comment]
great, was unable to reproduce on 1.2.0, thanks for the quick reply and sorry for the noise.
